Update Express tutorial to v5 (#40247)
mdn/content: files/en-us/learn_web_development/extensions/server-side/express_nodejs/routes/index.md

We'll also use route parameters as discussed in the next section.

### Route parameters

Route parameters are _named URL segments_ used to capture values at specific positions in the URL. The named segments are prefixed with a colon and then the name (E.g., `/:your_parameter_name/`). The captured values are stored in the `req.params` object using the parameter names as keys (E.g., `req.params.your_parameter_name`).

So for example, consider a URL encoded to contain information about users and books: `http://localhost:3000/users/34/books/8989`. We can extract this information as shown below, with the `userId` and `bookId` path parameters:

```js
app.get("/users/:userId/books/:bookId", (req, res) => {
  // Access userId via: req.params.userId
  // Access bookId via: req.params.bookId
  res.send(req.params);
});
```

> [!NOTE]
> The URL _/book/create_ will be matched by a route like `/book/:bookId` (because `:bookId` is a placeholder for _any_ string, therefore `create` matches). The first route that matches an incoming URL will be used, so if you want to process `/book/create` URLs specifically, their route handler must be defined before your `/book/:bookId` route.

> The names of route parameters can be any valid JavaScript identifier like `bookId` above, starting with a letter, `_` or `$`, digits after the first character, and no hyphens or spaces.

Reviewer: ```suggestion
Route parameter names (for example, `bookId`, above) can be any valid JavaScript identifier that starts with a letter, `_`, or `$`. You can include digits after the first character, but not hyphens and spaces.
```

You can also use names that aren't valid JavaScript identifiers, including spaces, hyphens, emoticons, or any other character, but you need to define them with a quoted string and access them using bracket notation.
For example:

```js
app.get('/users/:"user id"/books/:"book-id"', (req, res) => {
  // Access quoted param using bracket notation
  const user = req.params["user id"];
  const book = req.params["book-id"];
  res.send({ user, book });
});
```

### Wildcards

Wildcard parameters match one or more characters across multiple segments, returning each segment as a value in an array.
They are defined the same way as regular parameters, but are prefixed with an asterisk.

So for example, consider the URL `http://localhost:3000/users/34/books/8989`, we can extract all the information after `users/` with the `example` wildcard:

```js
app.get("/users/*example", (req, res) => {
  // req.params would contain { "example": ["34", "books", "8989"]}
  res.send(req.params);
});
```

### Optional parts

Braces can be used to define parts of the path that are optional.

> For for example, below we match a filename with any extension (or none).

Reviewer: ```suggestion
For example, below we match a filename with any extension (or none).
```

```js
app.get("/file/:filename{.:ext}", (req, res) => {
  // Given URL: http://localhost:3000/file/somefile.md`
  // req.params would contain { "filename": "somefile", "ext": "md"}
  res.send(req.params);
});
```

### Reserved characters

The following characters are reserved:`(()[]?+!)`.

> If you want to use them you will have to first escape them with the backslash (`\`).

Reviewer: ```suggestion
If you want to use them, you must escape them with a backslash (`\`).
```

You also can't use the pipe character (`|`) in a regular expression.

That's all you need to get started with routes.

> If needed you can find more information in the Express docs: [Basic routing](https://expressjs.com/en/starter/basic-routing.html) and [Routing guide](https://expressjs.com/en/guide/routing.html). The following sections show how we'll set up our routes and controllers for the LocalLibrary.

Reviewer: ```suggestion
If needed, you can find more information in the Express docs: [Basic routing](https://expressjs.com/en/starter/basic-routing.html) and [Routing guide](https://expressjs.com/en/guide/routing.html). The following sections show how we'll set up our routes and controllers for the LocalLibrary.
```

### Handling errors and exceptions in the route functions

The route functions shown earlier all have arguments `req` and `res`, which represent the request and response, respectively.

> Route functions are also called with a third argument `next`, which can be called to pass any errors or exceptions to the Express middleware chain, where they will eventually propagate to your global error handling code.

Reviewer: ```suggestion
Route functions are also passed a third argument, `next`, which contains a callback function that can be called to pass any errors or exceptions to the Express middleware chain, where they will eventually propagate to your global error handling code.
```

> From Express 5, `next` is called automatically with the rejection value if a route handler returns [Promise](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that subsequently rejects.

Reviewer: ```suggestion
From Express 5, `next` is called automatically with the rejection value if a route handler returns a [Promise](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that subsequently rejects; therefore, no application code is required for error handling.
```

Reviewer: I'm suggesting adding this phrase here to make it a bit clearer. I missed this earlier and was wondering where the error handling code had gone ;-)

This leads to very compact code when working with asynchronous promise-based APIs, in particular when using [`async` and `await`](/en-US/docs/Learn_web_development/Extensions/Async_JS/Promises#async_and_await).

For example, the following code uses the `find()` method to query a database and then renders the result.

> No application code is required for error handling.

Reviewer: Delete this line if you agree with my previous statement, or maybe add something that describes how you would then pick up the error that is automatically sent?

Author: ```suggestion
```

```js
exports.get("/about", async (req, res, next) => {
  const successfulResult = await About.find({}).exec();
  res.render("about_view", { title: "About", list: successfulResult });
});
```

The code below shows the same example using a promise chain.

> Note that if you wanted to, you could `catch` the error and implement your own custom handling.

Reviewer: ```suggestion
Note that if you wanted to, you could `catch()` the error and implement your own custom handling.
```

```js
exports.get("/about", (req, res, next) => {
  // Removed 'async'
  return About.find({})
    .exec()
    .then((successfulResult) => {
      res.render("about_view", { title: "About", list: successfulResult });
    });
  /*
    .catch(err => {
      next(err);
    });
    */
});
```

> [!NOTE]

> > Most modern APIs use asynchronous Promised-based APIs, so error handling is often that straightforward.

Reviewer: ```suggestion
> Most modern APIs are asynchronous and promise-based, so error handling is often that straightforward.
```

> Certainly that's all you really _need_ to know about error handling for this tutorial!

> Express 5 will automatically catch and forward exceptions that are thrown in synchronous code:

Reviewer: ```suggestion
Express 5 automatically catches and forwards exceptions that are thrown in synchronous code:
```

The previous phrasing made it sound like a future feature...but we are dealing with Express 5 now!

```js
app.get("/", (req, res) => {

>   throw new Error("SynchronousException"); // Express will catch this.

Reviewer: ```suggestion
  // Express will catch this
  throw new Error("SynchronousException");
```

Just to make the line shorter?

Author: Don't mind either way :-)

});
```

> But you must [`catch`](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) exceptions that occur in asynchronous code invoked by route handlers or middleware, as these will not be caught by the default code:

Reviewer: ```suggestion
However, you must [`catch`](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) exceptions occurring in asynchronous code invoked by route handlers or middleware. These will not be caught by the default code:
```

> In order for the framework to properly handle exceptions, they must be caught, and then forwarded as errors as shown in the previous section.

Author: Still working on this section. This module is no longer needed because Express 5 handles exceptions automatically. Just need to talk about it and link to docs.
